db/silkworm/stagedsync/stage_hashstate.cpp

```diff
@@ -323,15 +333,21 @@ void hashstate_unwind(mdbx::txn& txn, uint64_t unwind_to, HashstateOperation ope
     }
 
     (void)db::for_each(changeset_table, unwind_func);
-
 }
 
-StageResult unwind_hashstate(db::EnvConfig db_config, uint64_t unwind_to) {
+StageResult unwind_hashstate(db::EnvConfig db_config, uint64_t unwind_to, mdbx::txn* external_txn) {
```

Comment:
I'd have approached the matter in a different way.
Two different function signatures for function
* version A with db_config as parameter which eventually opens a transaction and calls B and commits on return)
* version B with txn as parameter which never commits

Higher functions either call A or B on behalf of how wide the transaction should be.
Also I'd change db_config to env

Reply:
I've reworked this PR. Now there's a new helper class, TransactionManager, that encapsulates the differences between the normal mode and the external transaction mode.